### PR TITLE
Add cachix github action

### LINKED ISSUE #1: Consider binary cache?
Thanks a lot for your great work! But it seems that it still takes some time for me to build the environment (~20mins) :(

Would you please consider providing a binary cache at, for example, cachix? That would be great!

There are ready-made GitHub Actions that generate the caches on each push, so it shouldn't be of too much a trouble.

### PR DESCRIPTION
**Pull request by SuperKenVery**
A few things:

1. Not sure why you used FHSenv, without it everything works but with it the shell integration of direnv is broken, so I removed it
2. Moved some `python3Packages.python-lsp-server` to `python3.withPackages`
3. Added `unrar`, `exploitdb`, `masscan` and `nmap`.
4. Added a github action config to automatically
    - Build devshell when push or PR occurs
    - Push them to cachix
    To use it, you would need to configure a cachix secret as described [here](https://nix.dev/guides/recipes/continuous-integration-github-actions)

### CODE REVIEW

**Comment by SoraTenshi** [comment]
The reason FHSEnv was part of this shell is because for binary exploitation challenges, you typicall have a shared object linked to the fhs path where dynamic libs (.so) reside.
That's obviously *muuuch* different on nixos, so i created the fhs environment in an effort to have this work.
I switched however over to use `pwninit` which is a fantastic tool that just patches the path where the library is expected. And that makes it work.

File: flake.nix [hunk 1]
```diff
@@ -18,44 +18,47 @@
         config.allowUnfree = true;
         config.allowUnsupportedSystem = true;
       };
-      ctf = pkgs.buildFHSEnv {
+      ctf = pkgs.mkShell {
         name = "ctf-env";
-        targetPkgs = pkgs: let
-          pythonStuff = pkgs.python311Full.withPackages (pp: [
+        buildInputs = let
+          pythonStuff = pkgs.python311.withPackages (pp: [
```
Comment (SoraTenshi) [inline flake.nix]:
```suggestion
          pythonStuff = pkgs.python312.withPackages (pp: [
```

I think we can safely update the python version

File: .direnv/bin/nix-direnv-reload [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+set -e
+if [[ ! -d "/home/ken/Codes/ctf" ]]; then
```
Comment (SoraTenshi) [inline .direnv/bin/nix-direnv-reload]: does the path need adjustments?
I have not a lot of experience with gh-actions or a like.

if it's something that should be executed locally, it'd probably make more sense to use the `$HOME` env variable

File: .github/workflows/build-and-cache.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    steps:
+      - uses: actions/checkout@v4
+      - uses: DeterminateSystems/nix-installer-action@main
+      - uses: DeterminateSystems/magic-nix-cache-action@main
+      - uses: DeterminateSystems/flake-checker-action@main
+      - uses: cachix/cachix-action@v14
+        with:
+          # https://nix.dev/guides/recipes/continuous-integration-github-actions
+          name: superkenvery
+          authToken: '${{ secrets.CACHIX_AUTH_TOKEN }}'
```
Comment (SoraTenshi) [inline .github/workflows/build-and-cache.yml:22]: I suppose i'd need to register on cachix?
probably also adjust based on that info. (e.g. add cachix token to the list of variables i guess)

File: .direnv/flake-inputs/01x5k4nlxcpyd85nnr0b9gm89rm8ff4x-source [hunk 4]
```diff
@@ -0,0 +1 @@
+/nix/store/01x5k4nlxcpyd85nnr0b9gm89rm8ff4x-source
```
Comment (SoraTenshi) [inline .direnv/flake-inputs/01x5k4nlxcpyd85nnr0b9gm89rm8ff4x-source]: also; what's with all those `.direnv` files?
Are they some sort of cachix thing?

Comment (SuperKenVery) [reply .direnv/flake-inputs/01x5k4nlxcpyd85nnr0b9gm89rm8ff4x-source]: > Are they some sort of cachix thing?

No, they are generated by `direnv` which enters an environment when you `cd` into a directory. In this case, we use nix dev shell.

I actually don't know whether I should commit this directory or not; some reddit posts suggest that I should. However, seeing that it contains my local directory, I think I should probably ignore those files from now on :D

File: .github/workflows/build-and-cache.yml (hunk 3 above)
Comment (SuperKenVery) [reply .github/workflows/build-and-cache.yml:22]: Yes you would need to register on cachix, if you want to use cachix as a binary cache.

There are also other options, including some self-host ones, if you don't like cachix. But AFAIK cachix is the most convenient and popular one.

Comment (SuperKenVery) [reply .github/workflows/build-and-cache.yml:22]: For information on how to setup the github action and get it running, [here](https://nix.dev/guides/recipes/continuous-integration-github-actions) is a good guide.

**Comment by SoraTenshi** [comment]
Thanks! I will as soon as i find time fix it to use cachix.

**Pull request merged**